codegen/src/main/twirl/templates/PlayScalaServer/Router.scala.txt

```diff
@@ -0,0 +1,49 @@
+@*
+ * Copyright (C) 2018 Lightbend Inc. <https://www.lightbend.com>
+ *@
+
+@(service: akka.grpc.gen.scaladsl.Service)
+
+package @service.packageName
+
+import javax.inject._
+
+import scala.concurrent.Future
+
+import akka.http.scaladsl.model.{HttpRequest, HttpResponse}
+import akka.stream.Materializer
+
+import play.api.mvc._
+import play.api.mvc.akkahttp.AkkaHttpHandler
+import play.api.routing.Router
+import play.api.routing.Router.Routes
+
+@@Singleton
+class @{service.name}Router @@Inject()(impl: controllers.@{service.name}Impl)(implicit mat: Materializer)
+  extends Router {
+
+  protected val prefix: String = @{service.name}.name
+
+  val handler = new AkkaHttpHandler {
+    val h = @{service.name}Handler(impl)
+    override def apply(request: HttpRequest): Future[HttpResponse] = h(request)
+  }
+
+  // could look at @{service.name}.name, but isn't this already covered in the route file?
+  override def routes: Routes = {
+    case rh if rh.path.startsWith(prefix) ⇒ handler
```

Comment:
If the `prefix` is set by the parent router then the prefix could be something like `/` and using `startsWith(prefix)` might not be precise enough. It might match paths that it shouldn't.

It might be safer to call `isDefinedAt` on the underlying `PartialFunction`. E.g.

```scala
  private val serviceFunc: PartialFunction[HttpRequest, Future[HttpResponse]] = GreeterServiceHandler(serviceImpl)(mat)

  private def toAkkaHttpRequest(rh: RequestHeader): HttpRequest = {
    rh.headers.asInstanceOf[AkkaHeadersWrapper].request
  }

  override def routes: Routes = new Routes {
    private val playHandler = AkkaHttpHandler(serviceFunc)
    override def isDefinedAt(rh: RequestHeader): Boolean = {
      val request = toAkkaHttpRequest(rh)
      val defined = serviceFunc.isDefinedAt(request)
      println(s"GreeterServiceRouter: route to ${rh.path} is $defined")
      defined
    }
    override def apply(rh: RequestHeader): Handler = playHandler
  }
```

Since this code uses some internal Play APIs (e.g. `asInstanceOf[AkkaHeadersWrapper]` it might make sense to add a helper Play to create a `Routes` given a `PartialFunction[HttpRequest, Future[HttpResponse]]`. This is similar to what @renatocaval suggested elsewhere.

Reply:
For gRPC specifically it seems uncommon to have other routes 'next to' the gRPC routes, so if the user specifies `/` as the prefix I think it might be reasonable to treat this as "you asked for it, you got it" there.

For the more general 'Akka HTTP routes in Play' case it might indeed be useful to really treat the route as a partial function all the way. Not sure about the performance impact. Might be OK to postpone that for now?